macro/R/forced_si.R

```diff
@@ -0,0 +1,162 @@
+#' @param former_state a list describing the individual's state before the transition
+#' @param new_state a list describing the individual's state after the transition
+#' @param curtime The time at which this transition fires.
+#' @return a list with the name and time.
+#' @export
+forced_si_observer <- function(transition_name, former_state, new_state, curtime) {
+  list(name = transition_name, curtime = curtime, id = former_state[["who"]])
+}
+
+
+#' Generate bites for a population of a forced_si model.
+#'
+#' @param people_cnt How many people there are
+#' @param bite_rate The number of bites per day.
+#' @param current_time The start time for bite events
+#' @param duration in days
+#' @return a list of arrays of bite times
+#' @example
+#' pop <- forced_si_population(5L, 0.4)
+#' pop[, "bites"] <- forced_si_create_bites(nrow(pop), 0.2, 0, 14)
+#'
+#' @export
+forced_si_create_bites <- function(people_cnt, bite_rate, current_time, duration) {
+  lapply(1:people_cnt, function(x) {
+    bites <- cumsum(rexp(round(bite_rate * duration * 3), bite_rate))
+    current_time + bites[bites < duration]
+  })
+}
+
+
+
+#' Creates a MASH module that runs forced SI.
+#'
+#' This module drives infection by supplying bites.
+#'
+#' @param parameters These are simulation parameters. They are
+#'     \code{recovery_rate} for recovery, \code{people_cnt} for
+#'     the number of individuals, \code{duration_days} for the number
+#'     of days per time step. If there are too few or too many parameters,
+#'     then this rejects the argument. We disallow extra parameters so that
+#'     misnamed parameters don't go unnoticed.
+#' @return A simulation object, which is a list.
+#' @export
+forced_si_module <- function(parameters) {
+  expected_parameters <- c("recovery_rate", "people_cnt", "duration_days")
+  stopifnot(all(names(parameters) %in% expected_parameters))
+  stopifnot(all(expected_parameters %in% names(parameters)))
+
+  transitions <- list(
+    infect = forced_si_infect(),
+    recover = forced_si_recover(parameters$recovery_rate)
+  )
+  people_cnt <- parameters$people_cnt
+  pfpr <- parameters$initial_pfpr
+  individuals <- forced_si_population(people_cnt, pfpr)
+
+  simulation <- continuous_simulation(
+    individuals,
+    transitions,
+    forced_si_observer
+  )
+  simulation[["parameters"]] <- parameters
+  simulation <- init_simulation(simulation)
+}
+
+
+#' Takes one time step of the discrete time step.
+#'
+#' @param simulation A forced-SI model.
+#' @param bites A list of bites for each person. Each item in the list is a vector
+#'     of times when an infectious bite happens.
+#' @return a simulation object
+#' @examples
+#' \dontrun{
+#' step_si_module(simulation, forced_si_create_bites(100, 1/20, current_time, 14))
+#' }
+#' @export
+step_si_module <- function(simulation, bites) {
+  duration_days <- simulation$parameters$duration_days
+  simulation$state[, "bites"] <- bites
```

Comment:
currently doesn't work, use `simulation$state$bites <- bites`

Reply:
I'm super confused b/c tests passed before I posted the PR. That's something I have to do before I post it, as a rule. I'm looking now...

Reply:
data.table_1.12.8 seems to allow this. It's described in the documentation. The `$bites` works too, though.

Reply:
Again, this runs but gives a warning message, which I didn't see before. Mind you, I just upgraded data.table yesterday. I'm suspicious. But I'm using your suggestion.